packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,10 +11,824 @@ import 'picker.dart';
+    // state.
+    if (mode == CupertinoDatePickerMode.time || mode == CupertinoDatePickerMode.dateAndTime)
+      return new _CupertinoDatePickerDateTimeState();
+    else
+      return new _CupertinoDatePickerDateState();
+  }
+}
+
+class _CupertinoDatePickerDateTimeState extends State<CupertinoDatePicker> {
+  int textDirectionFactor;
+  CupertinoLocalizations localizations;
+
+  // Alignment based on text direction. The variable name is self descriptive,
+  // however, when text direction is rtl, alignment is reversed.
+  Alignment alignCenterLeft;
+  Alignment alignCenterRight;
+
+  // The currently selected values of the date picker.
+  int selectedDayFromInitial; // The difference in days between the initial date and the currently selected date.
+  int selectedHour;
+  int selectedMinute;
+  int selectedAmPm; // 0 means AM, 1 means PM.
+
+  // The controller of the AM/PM column.
+  FixedExtentScrollController amPmController;
+
+  // Estimated width of columns.
+  final Map<int, double> estimatedColumnWidths = <int, double>{};
+
+  @override
+  void initState() {
+    super.initState();
+    selectedDayFromInitial = 0;
+    selectedHour = widget.initialDateTime.hour;
+    selectedMinute = widget.initialDateTime.minute;
+    selectedAmPm = 0;
+
+    if (!widget.use24hFormat) {
+      selectedAmPm = selectedHour ~/ 12;
+      selectedHour = selectedHour % 12;
+      if (selectedHour == 0)
+        selectedHour = 12;
+
+      amPmController = new FixedExtentScrollController(initialItem: selectedAmPm);
+    }
+  }
+
+  // Estimate the minimum width that each column needs to layout its content.
+  double _getColumnWidth(_pickerColumnType columnType) {
+    String longestText = '';
+
+    switch (columnType) {
+      case _pickerColumnType.date:
+        // Measuring the length of all possible date is impossible, so here
+        // just some dates are measured.
+        for (int i = 1; i <= 12; i++) {
+          final String date = localizations.datePickerMediumDate(
+            new DateTime(widget.initialDateTime.year, i, 25));
+          if (longestText.length < date.length)
+            longestText = date;
+        }
+        break;
+      case _pickerColumnType.hour:
+        for (int i = 0 ; i < 24; i++) {
+          final String hour = localizations.datePickerHour(i);
+          if (longestText.length < hour.length)
+            longestText = hour;
+        }
+        break;
+      case _pickerColumnType.minute:
+        for (int i = 0 ; i < 60; i++) {
+          final String minute = localizations.datePickerMinute(i);
+          if (longestText.length < minute.length)
+            longestText = minute;
+        }
+        break;
+      case _pickerColumnType.dayPeriod:
+        longestText =
+          localizations.anteMeridiemAbbreviation.length > localizations.postMeridiemAbbreviation.length
+            ? localizations.anteMeridiemAbbreviation
+            : localizations.postMeridiemAbbreviation;
+        break;
+      default:
```

Comment:
Don't add default here. It won't give an analyzer warning if we add new enum values. 

Just let the assert be outside the switch.